doc/design/drafts/dynamic_size_types.md

```diff
@@ -0,0 +1,562 @@
+# Dynamic Size Support for Zero Copy Types
+
+## Summary
+
+There are multiple obstacles when it comes to use true zero copy with types
+whose size is not known at compile time. At the moment a developer can pursue
+three approaches.
+
+1. Send via untyped publisher and provide the size at runtime to the publisher.
+   But defining message types with multiple dynamic sized containers would be
+   a real challenge when they should be constructed inside of this sample.
+
+2. One can perform a worst case estimation at compile time and declare
+   a type with container members which fullfil those estimations.
+   This could lead to a lot of wasted memory and if the worst case estimations
+   have to be adjusted all the applications which are using this type may have
+   to be adjusted as well since the capacity is part of the type.
```

Comment:
These estimations are also often hard to come by for more complicated structures, e.g. specific trees with internal meta data nodes etc. The estimates then are very pessimistic and as stated here lead to a lot of wasted memory which cannot be avoided by this approach. Hence it is not really an option if memory is a concern.